packages/svelte/src/internal/client/dom/blocks/boundary.js

```diff
@@ -84,37 +95,30 @@ export function boundary(node, props, boundary_fn) {
 
 				did_reset = true;
 
+				if (calling_on_error) {
+					w.reset_misuse();
+					throw error;
+				}
+
 				pause_effect(boundary_effect);
 
 				with_boundary(boundary, () => {
 					is_creating_fallback = false;
-					try {
-						boundary_effect = branch(() => boundary_fn(anchor));
-					} catch (error) {
-						// If the new subtree immediately throws during mount, warn the dev.
-						w.reset_misuse();
-						throw error;
-					}
+					boundary_effect = branch(() => boundary_fn(anchor));
 				});
 			};
 
 			var previous_reaction = active_reaction;
 
 			try {
 				set_active_reaction(null);
+				calling_on_error = true;
 				onerror?.(error, reset);
```

Comment:
@Rich-Harris People might want to auto-fix the state of the application inside the onerror for instance they might decide to show an error snackbar and reset a form without any confirmation from the user